File: running-app-backend/activity/serializers/group.py
from rest_framework import serializers

from activity.models import Group, Event

from account.serializers import UserSerializer
from activity.serializers.event import EventSerializer

class GroupSerializer(serializers.ModelSerializer):
    class Meta:
        model = Group
        fields = (
            "id",
            "name",
            "avatar",
            "total_distance",
            "total_duration"
        )
        extra_kwargs = {
            "id": {"read_only": True}
        }

class DetailGroupSerializer(serializers.ModelSerializer):
    total_distance = serializers.SerializerMethodField()
    number_of_participants = serializers.SerializerMethodField()
    total_duration = serializers.SerializerMethodField()
    rank = serializers.SerializerMethodField()
    event = EventSerializer()
    users = UserSerializer(many=True, read_only=True)

    def get_total_distance(self, instance):
        return instance.total_distance()
    
    def get_number_of_participants(self, instance):
        return instance.number_of_participants()
    
    def get_total_duration(self, instance):
        return instance.total_duration()

    def get_rank(self, instance):
        return instance.rank()
    
    class Meta:
        model = Group
        fields = "__all__"
        extra_kwargs = {
            "id": {"read_only": True}
        }
class CreateUpdateGroupSerializer(serializers.ModelSerializer):
    event_id = serializers.UUIDField()
    event = serializers.SerializerMethodField()

    def create(self, validated_data):
        event_id = validated_data.pop('event_id')
        return Group.objects.create(event_id=event_id, **validated_data)

    def get_event(self, instance):
        event = Event.objects.get(id=self.instance.event_id)
        return EventSerializer(event).data

    def to_representation(self, instance):
        instance = super().to_representation(instance)
        instance['event'] = self.get_event(instance)
        return instance

    class Meta:
        model = Group
        fields = (
            "id",
            "name", 
            "description",
            "avatar",
            "banner",
            "event_id",
            "event"
        )
        extra_kwargs = {
            "id": {"read_only": True}
        }

File: running_app/lib/models/activity/group.dart
import 'package:running_app/models/account/leaderboard.dart';

import '../account/user.dart';
import '../activity/event.dart';

class Group {
  final String? id;
  final String? name;
  final String? avatar;
  final int? totalDistance;
  final String? totalDuration;
  final int? totalParticipants;

  Group({
    required this.id,
    required this.name,
    required this.avatar,
    required this.totalDistance,
    required this.totalDuration,
    required this.totalParticipants,
  });

  Group.fromJson(Map<String, dynamic> json)
      : id = json['id'],
        name = json['name'],
        avatar = json['avatar'],
        totalParticipants = json['total_participants'],
        totalDistance = json['total_distance'],
        totalDuration = json['total_duration'];

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'avatar': avatar,
      'total_participants': totalParticipants,
      'total_distance': totalDistance,
      'total_duration': totalDuration,
    };
  }

  @override
  String toString() {
    return 'Group{id: $id, name: $name, avatar: $avatar, totalParticipants: $totalParticipants, totalDistance: $totalDistance, totalDuration: $totalDuration}';
  }
}

class DetailGroup extends Group {
  final int? rank;
  // final Event? event;
  final List<Leaderboard>? users;
  final String? description;

  DetailGroup({
    String? id,
    String? name,
    String? avatar,
    int? totalDistance,
    String? totalDuration,
    int? totalParticipants,
    required this.rank,
    // required this.event,
    required this.users,
    required this.description
  }) : super(
    id: id,
    name: name,
    avatar: avatar,
    totalParticipants: totalParticipants,
    totalDistance: totalDistance,
    totalDuration: totalDuration,
  );

  DetailGroup.fromJson(Map<String, dynamic> json)
      : rank = json['rank'],
        // event = Event.fromJson(json['event']),
        users = (json['users'] as List<dynamic>).map((user) => Leaderboard.fromJson(user)).toList(),
        description = json['description'],
        super.fromJson(json);

  @override
  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = super.toJson();
    data['rank'] = rank;
    // data['event'] = event?.toJson();
    data['users'] = users?.map((user) => user.toJson()).toList();
    data['description'] = description;
    return data;
  }

  @override
  String toString() {
    return 'DetailGroup{${super.toString()}, rank: $rank, event: , users: $users, description: $description}';
  }
}

class CreateGroup {
  final String? name;
  final String? description;
  final String? avatar;
  final String? banner;
  final String? eventId;

  CreateGroup({
    required this.name,
    required this.description,
    required this.eventId,
    this.avatar,
    this.banner,
  });

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'description': description,
      'avatar': avatar,
      'banner': banner,
      'event_id': eventId,
    };
  }

  @override
  String toString() {
    return '${toJson()}';
  }
}